test/jdk/java/util/Base64/TestEncodingDecodingLength.java

```diff
@@ -21,55 +21,91 @@
+    @Test
+    public void largeEncodeIAETest() throws IllegalAccessException,
+            InvocationTargetException, NoSuchMethodException {
+        Method m = getMethod(ENCODER,
+                "encodedOutLength", int.class, boolean.class);
+        // When throwOOME param is false, encodedOutLength should return -1 in
+        // this situation, which encode() uses to throw IAE
+        assertEquals(-1, m.invoke(ENCODER, LARGE_MEM_SIZE, false));
     }
 
-    private static final void checkOOM(String methodName, Runnable r) {
+    // Effectively tests that the overloaded encode() and encodeToString() methods
+    // throw OutOfMemoryError with array/buffer sized near Integer.MAX_VALUE
+    @Test
+    public void largeEncodeOOMETest() throws IllegalAccessException, NoSuchMethodException {
+        Method m = getMethod(ENCODER,
+                "encodedOutLength", int.class, boolean.class);
         try {
-            r.run();
-            throw new RuntimeException("OutOfMemoryError should have been thrown by: " + methodName);
-        } catch (OutOfMemoryError er) {}
+            m.invoke(ENCODER, LARGE_MEM_SIZE, true);
+        } catch (InvocationTargetException ex) {
+            Throwable rootEx = ex.getCause();
+            assertEquals(OutOfMemoryError.class, rootEx.getClass(), "00ME should be thrown");
+            assertEquals("Encoded size is too large", rootEx.getMessage());
+        }
     }
 
-    private static final void checkIAE(String methodName, Runnable r) {
+    // Effectively tests that the overloaded decode() methods do not throw
+    // OOME nor NASE with array/buffer sized near Integer.MAX_VALUE All the decode
+    // methods call decodedOutLength(), which is where the previously thrown
+    // OOME or NASE would occur at.
+    @Test
+    public void largeDecodeTest() throws IllegalAccessException, NoSuchMethodException {
+        Method m = getMethod(DECODER,
+                "decodedOutLength", byte[].class, int.class, int.class);
+        byte[] src = {1};
         try {
-            r.run();
-            throw new RuntimeException("IllegalArgumentException should have been thrown by: " + methodName);
-        } catch (IllegalArgumentException iae) {}
+            /*
+             decodedOutLength() takes the src array, position, and limit as params.
+             The src array will be indexed at limit-1 to search for padding.
+             To avoid passing an array with Integer.MAX_VALUE memory allocated, we
+             set position param to be -size. Since the initial length
+             is calculated as limit - position. This mocks the potential overflow
+             calculation and still allows the array to be indexed without an AIOBE.
+            */
+            m.invoke(DECODER, src, -LARGE_MEM_SIZE + 1, 1);
+        } catch (InvocationTargetException ex) {
+            fail("Decode should neither throw NASE or OOME: " + ex.getCause());
```

Comment:
Should we check the cause is either NASE or OOME here?

Reply:
The original test just validated there was no exception thrown, so I think we can just update the message to indicate an unexpected Exception.

I had added this as a comment but must have missed hitting the comment button

Reply:
Improved the error message to be more general, (as no exception should be thrown). Comments clarify that the method used to throw NASE and OOME.